### ISSUE TITLE
Feature request: Hot swap

### ISSUE THREAD

**Issue by dtrunk90** [pf4j/pf4j#267]
As discussed in #65 it would be nice to have a hot swap feature in order to replace plugins and let pf4j automagically stop and start/restart them.

**Comment by decebals** [comment]
@dtrunk90 
Thanks that you created a separate issue with this subject (hot swap). 
Someone explained in a [comment](https://github.com/pf4j/pf4j/issues/65#issuecomment-249180917) a possible scenario for hot swap/deployment in PF4J.
I explained in a [comment](https://github.com/pf4j/pf4j/issues/65#issuecomment-281335229) that the true hot swap is hard to achieve.
I implemented this concept in one of my project ([Pippo](https://github.com/pippo-java/pippo) a micro Java web framework). See [here](https://github.com/pippo-java/pippo/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+hot) the issue and [here](https://github.com/pippo-java/pippo/pull/355) the PR with some details about implementation. The bottom line is that the hot swap/deployment was implemented in Pippo restarting the web server (for example Jetty, Tomcat, Undertow, ...) when something was changed in Java sources of the project.
Why do you think that Eclipse, IDEA IntelliJ, Jenkins and other big applications (in Java) that are massively based on a concept like plugin/module/bundle, with a lot of resources in the back
resources (developers, money), require you a restart after you installed or updated a plugin/module? 
Furthermore, the Eclipse plugin system is based on OSGI, a more respectable, enterprise module system that somehow should solve this problem.

Sure, we can try together to see if we can find a solution. A simple implementation (POC) is to create a background task that checks the plugins directory for new, modified or removed files. When a modification is detected then PF4J, by default, will use methods from `PluginManager` like `loadPlugin` `startPlugin`, `stopPlugin`, `unloadPlugin` or `deletePlugin` to add, update or delete plugin.

Below links can be useful: 
- https://github.com/pf4j/pf4j/issues/226
- https://github.com/pf4j/pf4j/issues/217

**Comment by decebals** [comment]
@dtrunk90 
> The only project I know of which already has this feature and could be helpful to take a look into is 
Keycloak.

Do you talk about this [Keycloak](https://github.com/keycloak/keycloak)? Does this project come with a plugin system (embedded or not)? I will take a look if you point me to the right page.

**Comment by dtrunk90** [comment]
> Do you talk about this [Keycloak](https://github.com/keycloak/keycloak)? Does this project come with a plugin system (embedded or not)? I will take a look if you point me to the right page.

Yes, I'm talking about that Keycloak. The SPI is well documented here: https://www.keycloak.org/docs/latest/server_development/index.html#_providers

From [Section 6.2.1](https://www.keycloak.org/docs/latest/server_development/index.html#using-the-keycloak-deployer):
> If you copy your provider jar to the Keycloak standalone/deployments/ directory, your provider will automatically be deployed. Hot deployment works too.

I don't know if it's a Keycloak or WildFly feature.

**Comment by decebals** [comment]
> I don't know if it's a Keycloak or WildFly feature.

It's a WildFly feature. In PF4J the complexity is high because you have plugin with dependencies. We will see.

Like a detail, Jose (@jgomer2001) succeeded with [this](https://github.com/pf4j/pf4j/issues/217#issuecomment-392909053) approach.

**Comment by slovdahl** [comment]
> A simple implementation (POC) is to create a background task that checks the plugins directory for new, modified or removed plugins (zip or jar files). When a modification is detected then PF4J, by default, will use methods from PluginManager like loadPlugin startPlugin, stopPlugin, unloadPlugin or deletePlugin to add, update or delete plugin.

We have implemented something like this and used it for about 5 years already (using version 0.12.0). Until now we have not used dependencies between modules, but that has now become relevant.

When a plugin that is depended on by other plugins is reloaded, PF4J automatically stops and unloads the plugins that depend on the plugin being reloaded. This is all fine, but when the plugin is loaded and started again after that, we would like to automatically start the plugin that depend on the one being updated and that were loaded and running before the update. This is where I think PF4J could provide a little bit more help. It's possible to determine which plugins an update will unload and stop before it's done by using the `protected DependencyResolver dependencyResolver` field from `AbstractPluginManager`, and manually load and start those afterwards. I'm not sure how to do it though. In an ideal world, PF4J would be able to automagically handle it the same it stops and unloads dependent plugins, but that would require it to keep track of intermediate state.

Has anyone else had or solved the same kind of problem?

**Comment by fendo8888** [comment]
https://github.com/dcevm/dcevm  Based on whether this can be achieved？

**Comment by decebals** [comment]
Coming back to this one.

Two different things live under this title and I want to separate them.

The first is hot swap in the strict sense, replacing classes in a running JVM while the application still holds references to the old ones. My answer from 2019 has not changed and the ecosystem has not moved either: Eclipse, IntelliJ and Jenkins still ask for a restart after a plugin is installed or updated, with far more resources behind them than this project has. The dcevm link above shows how far this is from a library, it is a JVM patch. I don't plan to do this in PF4J, and I would rather say it once than leave the expectation open.

The second is what @slovdahl described in 2021: when a plugin is replaced, PF4J stops and unloads the plugins that depend on it and then leaves them there. That part is real, and it is not limited to hand written watchers. `UpdateManager.updatePlugin` in pf4j-update calls `deletePlugin` and then loads and starts the new version, and `deletePlugin` cascades over the dependents, so updating a plugin that others depend on leaves them stopped and unloaded while the method reports success.

I looked at having PF4J restart those dependents by itself and I don't want it, neither in the core nor in pf4j-update. A dependent restarted in place looks recovered, but the application may still hold extension instances from the class loader that was closed, which is the same "works most of the time, then becomes unpredictable" situation as above, only in a smaller box. What I want at that boundary is honesty: an update that takes dependents down should say so, or refuse and let the application decide, which is exactly what the applications named above do.

So the answer for this issue is no, both for hot swap and for restarting dependents automatically. I am converting it to a Discussion so the links and the context collected here stay available. The reporting side of `updatePlugin` is a separate and much smaller thing, and it belongs in pf4j-update.
